### PR TITLE
drivers: display: st7735r: Add rgb-is-inverted property

### PR DESCRIPTION
**Pull request by soburi**
Purpose
----------------------------------

This patch intends to support Longan Nano's built-in LCD module that uses ST7735S. (Adding workaround for  Longan Nano built-in LCD module's bug.)
(#44796 is PR to support Longan Nano's built-in LCD module, the PR is depends to this) 

Problem
---------------

Longan-Nano LCD module has an ST7735S controller that can drive with the ST7735R driver.
This LED module seems to be inverted pixel format between the MADCTL setting and the actual displaying color.

If pixel-format setting as RGB in MADCTL settings,
This driver 0xF100 displays as RED, 0x01F displays as GREEN in the usual case.
But this module is not so.
0xF100 displays as GREEN, 0x01F displays as RED.

I created a pattern for color confirmation with such a code and confirmed the actual display of Longan Nano. 

```
        for(int xx=0; xx<16; xx++) {
                uint16_t* buf16 = (uint16_t*)buf;
                for(int j=0; j<capabilities.x_resolution/16*capabilities.y_resolution; j++) {
                        uint16_t col = (1U << xx);
                        *buf16++ = (col&0xFF) << 8 | (col&0xFF00)>>8;
                }
                display_write(display_dev, (capabilities.x_resolution/16)*(15-xx), 0, &buf_desc, buf);
        }
```
We can see the bit and the corresponding color in this pattern.

MADCTL is 112 (order is RGB)

IMG-3592

the pixel format is 🟦🟦🟦🟦🟦🟩🟩🟩🟩🟩🟩🟥🟥🟥🟥🟥

(white square indicate (0,0))

MADCTL is 120 (order is BGR)

IMG-3591

the pixel format is 🟥🟥🟥🟥🟥🟩🟩🟩🟩🟩🟩🟦🟦🟦🟦🟦

The pixel format settings of the controller and the actual display are mismatched.

This behavior was observed in some samples from the module vendor.

In https://github.com/sipeed/Longan_GD32VF_examples case,

https://github.com/sipeed/Longan_GD32VF_examples/blob/7fe21406e73dbe49e9aa3c5a9cd24a44af3b5f41/gd32v_lcd/src/lcd/lcd.c#L330-L334
```
	LCD_WR_REG(0x36);
	if(USE_HORIZONTAL==0)LCD_WR_DATA8(0x08);
	else if(USE_HORIZONTAL==1)LCD_WR_DATA8(0xC8);
	else if(USE_HORIZONTAL==2)LCD_WR_DATA8(0x78);
	else LCD_WR_DATA8(0xA8);
```

Set 0x8(BGR) to MADCTL,

https://github.com/sipeed/Longan_GD32VF_examples/blob/7fe21406e73dbe49e9aa3c5a9cd24a44af3b5f41/gd32v_lcd/include/lcd/lcd.h#L117
```
#define RED           	 0xF800
```
But RED is define in RGB in format.

(I'm guessing there's a problem connecting the st7735 controller to the LCD panel.)

Of course, the MADCTL setting can swap RGB/BGR setting,
But that cannot solve all problem.

display_get_capabilities() API can retreive Display's pixel format.

ST7735R driver reports pixel-format of itself from MADCTL(36h) register.

Pixel-format that retrieve from display_get_capabilities() is same as MADCTL setting. That is different from the actual display pixel format in the Longan Nano LCD module's case.

As a result, correctly implemented applications that use display_get_capabilities() API will not display the correct color.
(such as samples/drivers/display.)

Workaround
------------------------

The problem is "Mismatching between reported pixel-format from display_get_capabilities() API and actual pixel-format for displaying."

Inverting pixel-format (swap RGB/BGR setting from MADCTL's register)  reported from display_get_capabilities() API can solve this mismatching.
(The primary problem is on the module side. But no way to fix MADCTL set pixel format and actual display color. It may be possible even if disassembling and rewiring the module.)

Typical LCD module behavior using ST7735R:

|Color data   | Reported pixel-format from display_get_capabilities()      | pixel-format set in MADCTL register        | Displayed color |
|-------------|----------------------|---------------|---------------------|
|0xF800       | PIXEL_FORMAT_RGB_565 | RGB           | RED                 | 
|0x07E0       | PIXEL_FORMAT_RGB_565 | RGB           | GREEN               | 
|0x001F       | PIXEL_FORMAT_RGB_565 | RGB           | BLUE                | 
|0xF800       | PIXEL_FORMAT_BGR_565 | BGR           | BLUE                | 
|0x07E0       | PIXEL_FORMAT_BGR_565 | BGR           | GREEN               | 
|0x001F       | PIXEL_FORMAT_BGR_565 | BGR           | RED                 | 

Longan Nano built-in LCD behavior:

|Color data   | Reported pixel-format from display_get_capabilities()      | pixel-format set in MADCTL register        | Displayed color |
|-------------|----------------------|---------------|---------------------|
|0xF800       | PIXEL_FORMAT_RGB_565 | RGB           | BLUE               | 
|0x07E0       | PIXEL_FORMAT_RGB_565 | RGB           | GREEN              | 
|0x001F       | PIXEL_FORMAT_RGB_565 | RGB           | RED                | 
|0xF800       | PIXEL_FORMAT_BGR_565 | BGR           | RED                | 
|0x07E0       | PIXEL_FORMAT_BGR_565 | BGR           | GREEN              | 
|0x001F       | PIXEL_FORMAT_BGR_565 | BGR           | BLUE               | 

workaround (Swap the RGB/BGR of the MADCTL registers and report them as capabilities):

|Color data   | Reported pixel-format from display_get_capabilities()      | pixel-format set in MADCTL register        | Displayed color |
|-------------|----------------------|---------------|---------------------|
|0xF800       | PIXEL_FORMAT_BGR_565 | RGB           | BLUE               | 
|0x07E0       | PIXEL_FORMAT_BGR_565 | RGB           | GREEN              | 
|0x001F       | PIXEL_FORMAT_BGR_565 | RGB           | RED                | 
|0xF800       | PIXEL_FORMAT_RGB_565 | BGR           | RED                | 
|0x07E0       | PIXEL_FORMAT_RGB_565 | BGR           | GREEN              | 
|0x001F       | PIXEL_FORMAT_RGB_565 | BGR           | BLUE               |

### CODE REVIEW

File: drivers/display/display_st7735r.c
Comment (jfischer-no) [inline drivers/display/display_st7735r.c:272]: Why is there cast to `(bool)`?

File: drivers/display/display_st7735r.c
Comment (jfischer-no) [inline drivers/display/display_st7735r.c:549]: What is the purpose of swap-color-order property since it does not change anything?

**Review by jfischer-no — CHANGES_REQUESTED** [review]

**Comment by ck-telecom** [comment]
wounderful!

File: drivers/display/display_st7735r.c
Comment (soburi) [reply drivers/display/display_st7735r.c]: XOR is bitwise operator, it cannot calculate correctly without cast.
It was not good notation to understanding.
Rewrited it with NOT operator.

File: drivers/display/display_st7735r.c
Comment (jfischer-no) [reply drivers/display/display_st7735r.c:549]: New property does not change the behavior of the controller, therefore it makes no sense anyway. Why can you not set correct value to `madctl` property (RGB-BGR ORDER)?

**Review by jfischer-no — CHANGES_REQUESTED** [review]

**Comment by ck-telecom** [comment]
how about change dt prop ```madctl```?

**Comment by soburi** [comment]
@jfischer-no , @ck-telecom 

> New property does not change the behavior of the controller, therefore it makes no sense anyway. Why can you not set correct value to madctl property (RGB-BGR ORDER)?

Shortly say, to work around bugs in the LCD module.

The built-in LCD module of Longan Nano that I'm trying to support with #44796 has a problem that is RGB/BGR setting is not match to actual displaying.

I created a pattern for color confirmation with such a code and confirmed the actual display of Longan Nano. 

```
        for(int xx=0; xx<16; xx++) {
                uint16_t* buf16 = (uint16_t*)buf;
                for(int j=0; j<capabilities.x_resolution/16*capabilities.y_resolution; j++) {
                        uint16_t col = (1U << xx);
                        *buf16++ = (col&0xFF) << 8 | (col&0xFF00)>>8;
                }
                display_write(display_dev, (capabilities.x_resolution/16)*(15-xx), 0, &buf_desc, buf);
        }
```
We can see the bit and the corresponding color in this pattern.

MADCTL is 112 (order is RGB)

IMG-3592

the pixel format is 🟦🟦🟦🟦🟦🟩🟩🟩🟩🟩🟩🟥🟥🟥🟥🟥

(white square indicate (0,0))

MADCTL is 120 (order is BGR)

IMG-3591

the pixel format is 🟥🟥🟥🟥🟥🟩🟩🟩🟩🟩🟩🟦🟦🟦🟦🟦

The pixel format settings of the controller and the actual display are mismatched.
(I'm guessing there's a problem connecting the st7735 controller to the LCD panel.)

We can draw correctly by fitting the pixel format to match the actual display, but there is one problem here.

The st7735r driver gets the pixel format of the display from the MADCTL settings. On the Longan Nano LCD, the RGB/BGR settings of the controller and displaying of that are different. So the correct pixel format cannot obtain from the capabilities structure.

By swapping the RGB/BGR order of MADCTL and the RGB/BGR order that can obtain by API, that can be satisfied the API specifications and the actual displaying.

The above is the purpose of introducing this property.

File: drivers/display/display_st7735r.c
Comment (soburi) [reply drivers/display/display_st7735r.c:549]: @jfischer-no , @ck-telecom 

> New property does not change the behavior of the controller, therefore it makes no sense anyway. Why can you not set correct value to madctl property (RGB-BGR ORDER)?

Shortly say, to work around bugs in the LCD module.

The built-in LCD module of Longan Nano that I'm trying to support with #44796 has a problem that is RGB/BGR setting is not match to actual displaying.

I created a pattern for color confirmation with such a code and confirmed the actual display of Longan Nano. 

```
        for(int xx=0; xx<16; xx++) {
                uint16_t* buf16 = (uint16_t*)buf;
                for(int j=0; j<capabilities.x_resolution/16*capabilities.y_resolution; j++) {
                        uint16_t col = (1U << xx);
                        *buf16++ = (col&0xFF) << 8 | (col&0xFF00)>>8;
                }
                display_write(display_dev, (capabilities.x_resolution/16)*(15-xx), 0, &buf_desc, buf);
        }
```
We can see the bit and the corresponding color in this pattern.

MADCTL is 112 (order is RGB)

IMG-3592

the pixel format is 🟦🟦🟦🟦🟦🟩🟩🟩🟩🟩🟩🟥🟥🟥🟥🟥

(white square indicate (0,0))

MADCTL is 120 (order is BGR)

IMG-3591

the pixel format is 🟥🟥🟥🟥🟥🟩🟩🟩🟩🟩🟩🟦🟦🟦🟦🟦

The pixel format settings of the controller and the actual display are mismatched.
(I'm guessing there's a problem connecting the st7735 controller to the LCD panel.)

We can draw correctly by fitting the pixel format to match the actual display, but there is one problem here.

The st7735r driver gets the pixel format of the display from the MADCTL settings. On the Longan Nano LCD, the RGB/BGR settings of the controller and displaying of that are different. So the correct pixel format cannot obtain from the capabilities structure.

By swapping the RGB/BGR order of MADCTL and the RGB/BGR order that can obtain by API, that can be satisfied the API specifications and the actual displaying.

The above is the purpose of introducing this property.

**Review by gmarull — CHANGES_REQUESTED** [review]
The order can be swapped by adjusting `madctl` in DT for this particular display, so adding a new property like this is redundant.

File: dts/bindings/display/sitronix,st7735r.yaml
Comment (gmarull) [inline dts/bindings/display/sitronix,st7735r.yaml:119]: If this is trying to fix an LCD issue (it works in RGB when instructed to do BGR and viceversa), the name of this property is misleading. Maybe `rgb-is-inverted`, and indicate clearly that this can be used for buggy displays.

Comment (soburi) [reply dts/bindings/display/sitronix,st7735r.yaml]: Thanks for the proposal. Apply it and update the description.

File: drivers/display/display_st7735r.c
Comment (gmarull) [inline drivers/display/display_st7735r.c:272]:
```suggestion
	if ((config->madctl & ST7735R_MADCTL_BGR) && !config->rgb_is_inverted) {
```

Comment (soburi) [reply drivers/display/display_st7735r.c:272]: Under the condition, the result is as follows.
It has bit problem.

|  MADCTL   |  inverted   |  result   |
|----|----|---|
|RGB| not inverted | RGB (false)|
|RGB| inverted | *RGB (false)* |
|BGR| not inverted | BGR (true)|
|BGR| inverted | RGB (false)|

The original expression is essentially XOR.

**Comment by soburi** [comment]
Update commit message and PR title.

File: drivers/display/display_st7735r.c
Comment (soburi) [reply drivers/display/display_st7735r.c:272]: @jfischer-no 
Thanks for comment.
I was not careful enough. I will be careful next time.

Comment (soburi) [reply drivers/display/display_st7735r.c:281]: I added a comment because it was difficult to understand and unkindly.

File: drivers/display/display_st7735r.c
Comment (jfischer-no) [inline drivers/display/display_st7735r.c:279]: I think it should be described differently. Do you have a link to the display's errata?

Comment (soburi) [reply drivers/display/display_st7735r.c:279]: @jfischer-no 
Sorry for my late response. 

I looked it up, but there was no mention of the errata.

#44796 supports the LCD module that built-in Longan Nano Dev-Board.

Dev-Board Manufacturer(sipeed.com) provides these documents, 

Dev-Board datasheet
https://dl.sipeed.com/fileList/LONGAN/Nano/Spec/Sipeed%20longan%20nano%20Datasheet%20V1.0.pdf

LCD module datasheet
https://dl.sipeed.com/fileList/LONGAN/Nano/HDK/0.96%20inch%2080x160%20IPS%20LCD.pdf

But, no mention is in these documents about errata.

LCD module manufacturer's page also has no good information about the module.

https://www.swicn.com/product-by-category/6

(I feel the manufacturer considers it is internal specs, not a bug.
They did not mention about ST7735 settings.)

**Comment by soburi** [comment]
Rebase to main.

**Comment by mbolivar-nordic** [comment]
@jfischer-no ping

File: drivers/display/display_st7735r.c
Comment (jfischer-no) [reply drivers/display/display_st7735r.c:280]: @mbolivar-nordic @soburi Would it not be better to handle it using specific compatible? Something like
`
compatible: "longan-nano,st7735r"`

```
#if DT_HAS_COMPAT_STATUS_OKAY(longan_nano_st7735r)
     apply_fix_for_longan_nano();
#endif
```

Comment (soburi) [reply drivers/display/display_st7735r.c:280]: @jfischer-no 
I agree this workaround enable only in the case the by by specific compatible.
I'll try rework it.

Comment (jfischer-no) [reply drivers/display/display_st7735r.c:280]: Please wait for @mbolivar-nordic feedback :-)

Comment (soburi) [reply drivers/display/display_st7735r.c:280]: I noticed a problem.
before comma part of compatible name must be a vender prefix... :-(

Comment (jfischer-no) [reply drivers/display/display_st7735r.c:280]: There is dts/bindings/vendor-prefixes.txt

Comment (soburi) [reply drivers/display/display_st7735r.c:280]: I think it is not good idea. longan_nano is board name. I think should not be in vender-prefixs.txt
If replace longan_nano by that manufacturer name(sipeed,st7735r.yaml), 
this look like all of their product that use st7735r may affect this workaround... :-(

Comment (jfischer-no) [reply drivers/display/display_st7735r.c:280]: longan-nano was just an example

Comment (soburi) [reply drivers/display/display_st7735r.c:280]: I can't think good alternative yet.
I also wait @mbolivar-nordic feedback.

Comment (mbolivar-nordic) [reply drivers/display/display_st7735r.c:280]: Thank you for your detailed description of the problem @soburi. It seems like we are dealing with a device "quirk". I think a simple property is probably OK if a simple "inverted" is the only quirk you ever expect to see for a display handled by this controller.

However, I wonder if you might have other problems with other vendors. For example, if MADCTL says the color order is RGB, but the actual display order from some vendor is GRB. If that happens, `rgb-is-inverted` will not be enough. Do you think this is possible?

If so, you might want to define a `quirky-display-color-order` property that lets you define the actual display color order instead. For my example, you could write something like `quirky-display-color-order = <GREEN>, <RED>, <BLUE>;`. 

This is basically what we have to do in LED and LED strip bindings like this one:

https://docs.zephyrproject.org/latest/build/dts/api/bindings/led_strip/worldsemi%2Cws2812-spi.html#dtbinding-worldsemi-ws2812-spi

See the `color-mapping` property for details and the `drivers/led_strip/ws2812_spi.c` driver for how it's handled for that device.

In summary, if you think the current `rgb-is-inverted` property will be general enough to handle all vendor quirks, then it is OK for me. If you think vendors will be "creative" and pick any color order sometimes, you may want to consider a more general property.

I leave it to you to decide; either approach seems fine to me from the DTS side.

Comment (soburi) [reply drivers/display/display_st7735r.c:280]: Hi, @mbolivar-nordic,

> For example, if MADCTL says the color order is RGB, but the actual display order from some vendor is GRB. If that happens, rgb-is-inverted will not be enough. Do you think this is possible?

I think it cannot. MADCTL is binary property. It can select only from RGB or BGR format.

By the way, ST7735 also has an 'RGB' inverting pin' that control by wire.
From it, We can say that ST7735 considers RGB/BGR inverting but not considering other pixel formats. (Of course, It does not mention other pixel formats in datasheets.)

So I think 'rgb-is-inverted' property is enough to support functions of ST7735.

Comment (mbolivar-nordic) [reply drivers/display/display_st7735r.c:280]: > So I think 'rgb-is-inverted' property is enough to support functions of ST7735.

Very well, that sounds good to me as well. Thank you for the explanation.

**Review by jfischer-no — DISMISSED** [review]

**Comment by soburi** [comment]
@jfischer-no, @mbolivar-nordic
This PR was marked as staled, and I rebased it to the newest. (no modification, just only rebasing)
Could you re-check and approve if you may ok?

@gmarull, @galak
Could you review this PR if you have a time?

https://github.com/zephyrproject-rtos/zephyr/pull/44795#pullrequestreview-943841317

I think the following conversation will give you the answer.

https://github.com/zephyrproject-rtos/zephyr/pull/44795#discussion_r970158759

Summary of conversation:
- The Longan-Nano built-in LED is a "quirky" device.
- This device has a "quirk" in MADCTL that inverted the color ordering from the settings.
- Since MADCTL is a bool property, adding the `rgb-is-inverted` property can handle the same problem.

**Review by jfischer-no — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
@gmarull this is blocked on your NAK

**Review by gmarull — APPROVED** [review]
lgtm, sorry for the delay (PR got buried in my inbox)

**Pull request merged**
